Change 'docker run' exit codes to distinguish docker/contained errors

sallyom opened the pull request:
close #6734

The purpose of this PR is for users to distinguish Docker errors from
contained command errors.
This PR modifies 'docker run' exit codes to follow the chroot standard
for exit codes.
Exit status:
125 if 'docker run' itself fails
126 if contained command cannot be invoked
127 if contained command cannot be found
the exit status otherwise

Signed-off-by: Sally O'Malley somalley@redhat.com

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

sallyom commented:
I am an intern working with @rhatdan.

rhatdan commented:
We are basing this off of the "defacto" chroot standard.  Bash has some rules around this also.

LK4D4 commented:
looks ok for me.
ping @docker/core-maintainers

tiborvass commented:
@sallyom @LK4D4 Also see #13907 that just got merged

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "exitCodes" git@github.com:sallyom/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

sallyom commented:
Any word on this?

rhatdan commented:
@crosbymichael @jfrazelle @tianon

crosbymichael commented:
I think this is a good idea.  Do you think we would break anyone relying on the existing error codes or do you think they are just checking for a non-zero return?

rhatdan commented:
I would doubt anyone would check on this, but if they were that would be a bug in their code.  Currently you can  not differentiate between docker errors and errors in apps in containers, so looking for exit code of 1 makes little sense.

sallyom commented:
will this be considered before Docker 1.8?

rhatdan commented:
@crosbymichael @jfrazelle @tianon @cpuguy83

cpuguy83 commented:
+1 makes sense. Moving to code review since @crosbymichael seemed to be ok with it as well.

cpuguy83 commented on api/client/run.go:
```diff
@@ -35,6 +38,25 @@ func (cid *cidFile) Write(id string) error {
+// Exit code 125 signals docker failure
+// Exit code 126 signals contained cmd not executable
+// Exit code 127 for contained cmd not found
+func RunError(err error) error {
```
Does this need to be exported?

cpuguy83 commented:
Instead of having CLI command functions call `RunError`, call it from https://github.com/docker/docker/blob/master/api/client/cli.go#L90 and the command functions can still just `return err`.

duglin commented:
should at least some of these codes make it into the container's metadata so that we see them via an inspect() ?

cpuguy83 commented on api/client/run.go:
```diff
@@ -35,6 +38,25 @@ func (cid *cidFile) Write(id string) error {
+// Exit code 127 for contained cmd not found
+func RunError(err error) error {
+	var statusError StatusError
+	if strings.Contains(err.Error(), syscall.ENOENT.Error()) {
```
I think a switch is better for this.

rhatdan commented:
@cpuguy83 I had @sallyom simplify runError to only use three if statements.  Not sure why switch would improve this.

We only want runError to be called on Run and Exec Commands.  Wouldn't moving this runError to cli.go cause it to be called on other cli commands potentially causing additional errors? 

Also moving to cli.go, could cause error codes returned by the command executed within the container to get eaten.

tiborvass commented:
@sallyom rebase please

sallyom committed:
Change 'docker run' exit codes to distinguish docker/contained errors

The purpose of this PR is for users to distinguish Docker errors from
contained command errors.
This PR modifies 'docker run' exit codes to follow the chroot standard
for exit codes.
Exit status:
125 if 'docker run' itself fails
126 if contained command cannot be invoked
127 if contained command cannot be found
the exit status otherwise

Signed-off-by: Sally O'Malley <somalley@redhat.com>

thaJeztah commented:
I see this was rebased

ping @cpuguy83 @duglin PTAL

duglin commented:
@sallyom any comment on this https://github.com/docker/docker/pull/14012#issuecomment-122396476 ?

sallyom commented:
@duglin rebased, PTAL thanks

cpuguy83 commented:
Tests are failing against Windows.

LK4D4 commented:
@sallyom Yup, indeed windows tests are failing. Maybe @jhowardmsft can help

jhowardmsft commented:
I can't see anything obvious in here - maybe Windows golang doesn't allow exit codes that high for some reason? I can take a look in the morning - heading out of the office shortly.

duglin commented:
@sallyom https://github.com/docker/docker/pull/14012#issuecomment-122396476

For example, State.ExitCode is -1 instead of 127.  Should it be 127?

sallyom commented:
Thanks @jhowardmsft , I was going to ping you on irc for this!   I have been scouring the code to pinpoint the windows failure, too.  Looking forward to resolving this.

jhowardmsft commented:
OK, the problem is the string comparison. With the following little test program:

package main

import (
    "fmt"
    "os/exec"
    "syscall"
)

func main () {
    fmt.Println(syscall.ENOENT.Error())
    fmt.Println(exec.ErrNotFound.Error())
}

Windows output
Y:\test>test
The system cannot find the file specified.
executable file not found in %PATH%

Linux output
Y:\test>type a.a
no such file or directory
executable file not found in $PATH

sallyom commented:
Ok, @jhowardmsft thanks!  hopefully it will pass now.

sallyom commented:
Made a change to compare to Windows error msgs, still failing.  I should have time to look closer this afternoon.

duglin commented:
@sallyom I'd really appreciate a reply to https://github.com/docker/docker/pull/14012#issuecomment-129637212

sallyom commented:
@duglin 
https://github.com/docker/docker/blob/master/daemon/execdriver/windows/run.go#L236  - changed to:
return execdriver.ExitStatus{ExitCode: 127}, err    
This should be 127, not -1. 
I had this change already but windows test failed if I remember correctly?  Trying again...

duglin commented:
I probably wasn’t clear.  Unless I did something wrong, when I use your PR and try to start an exe that doesn’t exist in the container and then use “docker inspect” to ask for the exit code I see “-1” not “127”.  Should the new exit code values that you defined also be stored in the container’s ExitCode metadata?  It seems like it.


> On Aug 13, 2015, at 1:14 AM, sallyom <notifications@github.com> wrote:
> 
> @duglin <https://github.com/duglin> 
> https://github.com/docker/docker/blob/master/daemon/execdriver/windows/run.go#L236 <https://github.com/docker/docker/blob/master/daemon/execdriver/windows/run.go#L236> - changed to:
> return execdriver.ExitStatus{ExitCode: 127}, err
> 
> This should be 127, not -1. 
> I had this change already but windows test failed if I remember correctly? Trying again...
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/docker/docker/pull/14012#issuecomment-130377638>.
>

sallyom commented:
@duglin Yes, i see that now, thanks for clarifying.  Yes it should definitely be 127.  I'll make the fix.  Thanks for the help.

sallyom commented:
@duglin I made a fix so new exit code values get stored in the container's Exitcode metadata.  
126 for "permission denied" error, ex) docker run busybox /etc
127 for "executable file not found" or "no such file or directory"
  ex) docker run busybox foo, docker run busybox /bin/foo  
thanks again for the review.

sallyom commented:
@duglin non-Windows tests are clear after implementing your fb on the ExitCode metadata.  
@jhowardmsft tried adding Windows specific msgs to comparison, Windows still fails, still looking into it.

sallyom commented:
@duglin please confirm that ExitCode metadata is consistent now, thanks.
@jhowardmsft still looking at Windows failure, tried adjusting string comparison.  Looking closer today.

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2826,3 +2826,48 @@ func (s *DockerSuite) TestRunCreateContainerFailedCleanUp(c *check.C) {
+	out, exit, err := runCommandWithOutput(runCmd)
+	if !(err != nil && exit == 127) {
+		c.Fatalf("Run non-executable command should have errored with exit code 127, but we got out: %s, exit: %d, err: %s", out, exit, err)
+	}
```
Can you please add a check of the "docker inspect" output to make sure that State.ExitCode is also 127?  Just to be sure.

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2826,3 +2826,48 @@ func (s *DockerSuite) TestRunCreateContainerFailedCleanUp(c *check.C) {
+}
+
+// TestRunNonExistingCmd checks that 'docker run busybox /bin/foo' exits with code 127'
+func (s *DockerSuite) TestRunNonExistingCmd(c *check.C) {
```
How is this test different from the previous one?  In both cases we appear to be trying to run an exe that doesn't exit.

sallyom replied:
'run busybox foo' exits with syscall.ENOENT "no such file or directory" and 'run busybox /bin/foo' exits with exec.ErrNotFound "executable file not found in $PATH" 
(both error msgs are in the string compare that sets the exit code to 127, hence the test for each)

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2826,3 +2826,48 @@ func (s *DockerSuite) TestRunCreateContainerFailedCleanUp(c *check.C) {
+	out, exit, err := runCommandWithOutput(runCmd)
+	if !(err != nil && exit == 126) {
+		c.Fatalf("Run cmd that cannot be invoked should have errored with code 126, but we got out: %s, exit: %d, err: %s", out, exit, err)
+	}
```
Please check State.ExitCode here too.

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2826,3 +2826,48 @@ func (s *DockerSuite) TestRunCreateContainerFailedCleanUp(c *check.C) {
+	out, exit, err := runCommandWithOutput(runCmd)
+	if !(err != nil && exit == 125 && strings.Contains(out, "Unable to find image")) {
+		c.Fatalf("Run non-existing image should have errored with 'Unable to find image' code 125, but we got out: %s, exit: %d, err: %s", out, exit, err)
+	}
```
Check State.ExitCode.

duglin replied:
nvm, no container to check.  doi. :-)

duglin commented:
@sallyom yup - look like it works, thanks.  Just a few minor comments about the tests.

sallyom commented:
@LK4D4 @cpuguy83 
https://github.com/docker/docker/blob/master/integration-cli/utils.go#L31-#L40
This should be something other than 127, right?

bfirsh commented:
Thanks for your patience here, @sallyom! Any chance you could rebase and address @duglin's comments?

sallyom commented:
@bfirsh sorry for the delay, had a hard time getting the State.ExitCode tests to pass, strange errors in many tests similar to https://github.com/docker/docker/issues/15876 I believe.  Funny, my added tests were all passing but the rest of the suite blew up.  Finally passing (at least locally, tbd here).  Rebased, also.

sallyom commented:
also, wondering if https://github.com/docker/docker/blob/master/integration-cli/utils.go#L36:#L37 should be changed to exit code 125 rather than 127 for 'docker run' failure?

sallyom commented:
**As you can see from the windows build output, the State.ExitCode is set correctly.**
FAIL: docker_cli_run_test.go:2791: DockerSuite.TestRunNonExecutableCmd
docker_cli_run_test.go:2797:
    c.Fatalf("Run non-executable command should have errored with exit code 127, but we got exit: %d, State.ExitCode: %s", exit, stateExitCode)
... Error: Run non-executable command should have errored with exit code 127, but we got exit: 1, **State.ExitCode: 127**

@jhowardmsft  `docker inspect -f {{.State.ExitCode}} c.ID`in `daemon/execdriver/windows/exec.go` and in `daemon/execdriver/windows/run.go` to set exitCode == output from `docker inspect` should work?

jhowardmsft commented:
Should do. Just tried 
- docker run --rm windowsservercore cmd in first window
- docker ps in second window
- exit 83 in first windows
- docker inspect -f {{.State.ExitCode}} ac15 in second windows (ac15 was the ID)

And got 83 back as a result.

jhowardmsft commented:
Ah, but I don't think this will work with exit from a docker exec. Just docker run.

sallyom commented:
@jhowardmsft can you run `docker run --name test busybox foo`, check the exit code (it's probably 1)? and then `docker inspect -f {{.State.ExitCode}} test` and check the exit code (should be 127) . Please, thank you!
If the exit code of the docker run is 127, that's great, it means I have to change my test.  If not, I'm still on the task of figuring out where it's set, exactly, and where to set to 127.  We

sallyom commented:
@jhowardmsft please and thanks so much, can you run:
`docker run --name test` busybox foo` verify exits 1 (it will be great if it exits 127, then I'll rework the test).
`docker inspect -f {{.State.ExitCode}} test` verify it exits 127.
thanks again for your help.

jhowardmsft commented:
Well can't do busybox, but tried docker run -it windowsservercore doesnotexist, and inspect gave 127 on your branch. Not sure where I should be checking the exit code of 1 from though?

sallyom commented:
I'm looking for the exit code of the last command, like  `echo $?` in bash :)

sallyom commented:
@jhowardmsft @LK4D4 @cpuguy83 
(not sure if my windows "fix" is acceptable, PTAL, thanks)

jhowardmsft commented on integration-cli/docker_utils.go:
```diff
@@ -679,6 +679,22 @@ func dockerCmdInDirWithTimeout(timeout time.Duration, path string, args ...strin
+	if err != nil {
+		c.Fatal(err, out)
+	}
+	// return 127 for Windows build testing
```
Can you explain why this is Windows specific?

sallyom replied:
In Windows, in testing 'docker run busybox /bin/foo' or 'docker run busybox foo', container exits with a code of 1, but with a State.ExitCode of "127".  So if you do 'docker inspect' you'll see the ExitCode of "127" but checking the exit code of last cmd will give 1.  It appears to be quite tricky to get the container to exit with 127 when running in Windows, it exits with a 1, regardless of any changes I've made to the code.  So instead, in testing, I'm ensuring that the container exited with the State.Exit.Code of "127" and if   the exit code != 127 (this ONLY happens during Windows build) then just manually set the code to 127 for the test IF the proper error msg is in State.Error.  The test passes if a) State.ExitCode is "127"  and proper 127 error msg was found in State.Error (Windows) or b) State.ExitCode is "127" and exit code is 127. Implementing the changes in this PR will give users the benefit of more information about why the container exited with a non-zero code.

jhowardmsft replied:
Just to be clear - this is running from a Windows CLI against a Linux daemon when you say "In Windows...." and later "when running in Windows".

If so, the exit code is set by the daemon, so there shouldn't be a reason you have to do client-specific processing.

sallyom replied:
yes, understood. I see my wording is misleading, updating.  thanks.

jhowardmsft replied:
I'm still really confused here. I did not see a difference between a Linux CLI and a Windows CLI pointing to a Linux daemon. On both I ran docker run busybox foo, and both came back with State.ExitCode=-1 when run on master.

sallyom replied:
*  From master, I also get a State.ExitCode of -1.  However, I get output of **1** from `echo $?` 
*  From my branch, I get a State.ExitCode of 127 (Windows and Linux CLI) but with Windows CLI the  exit is still 1, while with Linux the exit is 127.  
*  What I really don't understand (yet) is why  with Windows CLI  exit code and State.ExitCode both are 126 in the case of `docker run busybox /etc' when the error is "permission denied" from contained cmd.

jhowardmsft replied:
No idea.... But FWIW, on master I see that Windows is perhaps already doing what you're wanting. docker run busybox foo, pointing to a Linux box. State.ExitCode=-1 on both platforms. echo %ERRORLEVEL% on Windows gives 127. But yes, echo $? on Linux is 1. 

Repeating same on master but using your /etc example. From a both CLI, State.ExitCode=-1. Linux $?=1, Windows %ERRORLEVEL%=1.

sallyom commented:
@LK4D4 @cpuguy83  thoughts on the Windows integration testing solution (please see my response above,  thanks :)

sallyom commented:
@cpuguy83  @LK4D4 @duglin @jhowardmsft @rhatdan 
Please take another look, up-to-date, all tests are passing now.  Thanks.

jhowardmsft commented on daemon/container.go:
```diff
@@ -262,9 +264,16 @@ func (container *Container) Start() (err error) {
 	defer func() {
 		if err != nil {
 			container.setError(err)
+			if strings.Contains(err.Error(), syscall.ENOENT.Error()) ||
```
Should this block not have the same checks as in runerror on the CLI side?

sallyom replied:
Added comment https://github.com/docker/docker/pull/14012/files#diff-07773483549b2d3c003c53614e1e2ab2R45 to clarify.

sallyom commented:
@cpuguy83 @LK4D4 @duglin @rhatdan up to date, tests passing, please take another look, thanks.

runcom commented on api/client/run.go:
```diff
@@ -36,6 +38,27 @@ func (cid *cidFile) Write(id string) error {
+		return Cli.StatusError{StatusCode: 126}
+	}
+	if strings.Contains(err.Error(), "latest not found") {
+		return Cli.StatusError{StatusCode: 125}
```
why don't these `StatusError` (125,126,127) contain a `Status` field set to `err.Error` (like the generic 125)?

now if you hit an error from 125 (latest not found), 126 or 127 the output is:
```
root@224662c1534f:/go/src/github.com/docker/docker# docker run busybox /etc
exec: "/etc": permission denied
```
while before this patch (and with `Status` field populated) output is:
```
root@224662c1534f:/go/src/github.com/docker/docker# docker run busybox /etc
exec: "/etc": permission denied
Error response from daemon: Cannot start container 62f29bd2a514c57fd0708b6a7dddf62008b1fc73a1b4797eb497e34f39b20b19: [8] System error: exec: "/etc": permission denied
```
is this change intented?

rhatdan commented:
Probably not intentional, but I like it from a usability point of view.

runcom commented on api/client/run.go:
```diff
@@ -36,6 +38,27 @@ func (cid *cidFile) Write(id string) error {
+	if strings.Contains(err.Error(), syscall.EACCES.Error()) {
+		return Cli.StatusError{StatusCode: 126}
+	}
+	if strings.Contains(err.Error(), "latest not found") {
```
```
# docker run busybox:notlatest true
Unable to find image 'busybox:notlatest' locally
Pulling repository docker.io/library/busybox
Tag notlatest not found in repository docker.io/library/busybox

# echo $?
1

# docker run cpuguy83/runcomtest          
Unable to find image 'cpuguy83/runcomtest:latest' locally
Pulling repository docker.io/cpuguy83/runcomtest

# echo $?
125
```
If I'm trying to pull an image by a not-latest tag the error is shown but the exit code is `1`
If I'm trying to pull an image looking for latest tag but it doesn't have it error isn't shown and exit code is `127`
instead of looking for `latest not found` should this look for `not found in repository` to be consistent for every tag? (not sure why you don't want to show the error if the tag isn't latest)

runcom commented:
I'm wondering if the introduction of the new `errors` pkg would help in this instead of just checking strings content. So the third case (125) could be just a type check on some error type from docker.
ping @duglin

cpuguy83 commented:
@runcom My thoughts as well.

sallyom commented:
@cpuguy83 @runcom I'm working on that right now, actually.  No need to make any changes in api/client/run.go....I'll ping you when I've got it cleaned up.  Thanks.

sallyom commented:
ping @cpuguy83 @runcom Thanks for taking another look at this.

thaJeztah commented:
@sallyom looks like this needs another rebase as well :cry:

sallyom commented:
rebased :)

runcom commented on daemon/container.go:
```diff
@@ -257,7 +257,7 @@ func (container *Container) Start() (err error) {
 			// if no one else has set it, make sure we don't leave it at zero
 			if container.ExitCode == 0 {
-				container.ExitCode = 128
+				container.ExitCode = 125
```
I'd rather leave 128 here just to be sure we know this is the case and we can differentiate from the other codes

rhatdan replied:
Is exit code 128 ever defined to mean anything in  similar commands like chroot or bash?  Googling it shows a hole mismatch of definitions. I have no problem leaving it but we need to document what it means so the qe people could test for this exit code and describe what i

runcom commented on daemon/start.go:
```diff
@@ -45,7 +47,20 @@ func (daemon *Daemon) ContainerStart(name string, hostConfig *runconfig.HostConf
 
 	if err := container.Start(); err != nil {
-		return derr.ErrorCodeCantStart.WithArgs(name, utils.GetErrorMessage(err))
+		derrErr := derr.ErrorCodeCantStart.WithArgs(name, utils.GetErrorMessage(err))
```
I don't really like this whole logic here, isn't it better to hide all of this inside `container.Start()`?

rhatdan replied:
Yes probably.  @sallyom See about moving this into container.Start()

sallyom replied:
np, I was building on what was there already but I agree it should be in container.Start()

runcom commented on pkg/mflag/flag.go:
```diff
@@ -1058,7 +1058,7 @@ func (fs *FlagSet) Parse(arguments []string) error {
 			return err
 		case ExitOnError:
-			os.Exit(2)
+			os.Exit(125)
```
mmm this sounds weird here in pkg/mflag

rhatdan replied:
Why?  125 means the command failed in a lot of situations (Or ECANCELLED in errno).  Exit 2 means either nothing (or file does not exist in ERRNO).

I think 125 is a more correct error.

runcom replied:
fair

sallyom replied:
working on feedback now, I've fixed the 'image doesn't not exist' errors by adding a similar check in api/client/run.go for the already existing derr.ErrorCodeNoSuchImageHash and derr.ErrorCodeNoSuchImageTag errors that were created in daemon/create.go --should have seen this before :(   working on other feedback now....

runcom commented on api/client/run.go:
```diff
@@ -115,6 +134,10 @@ func (cli *DockerCli) CmdRun(args ...string) error {
 
 	createResponse, err := cli.createContainer(config, hostConfig, hostConfig.ContainerIDFile, *flName)
 	if err != nil {
+		if strings.Contains(err.Error(), "Error: image library") {
```
this is wrong:
```
root@6ec0bedaeade:/go/src/github.com/docker/docker# docker run runcom/busybox true
Unable to find image 'runcom/busybox:latest' locally
Pulling repository docker.io/runcom/busybox
Error: image runcom/busybox:latest not found

root@6ec0bedaeade:/go/src/github.com/docker/docker# echo $?
1
```

runcom commented:
some other problems with latest tag:

```
root@6ec0bedaeade:/go/src/github.com/docker/docker# docker run busybox:notlatest true
Unable to find image 'busybox:notlatest' locally
Pulling repository docker.io/library/busybox
Tag notlatest not found in repository docker.io/library/busybox

# echo $?
1

root@6ec0bedaeade:/go/src/github.com/docker/docker# docker run cpuguy83/runcomtest
Unable to find image 'cpuguy83/runcomtest:latest' locally
Pulling repository docker.io/cpuguy83/runcomtest
Tag latest not found in repository docker.io/cpuguy83/runcomtest

root@6ec0bedaeade:/go/src/github.com/docker/docker# echo $?
1
```
- If I'm trying to pull an image by a not-latest tag, the error is 1 not 125
- If I'm trying to pull an image looking for latest tag but it doesn't have it, the error is 1 not 125

runcom commented:
Did a quick review and there still seems to be inconsistencies, will continue tomorrow
ping @cpuguy83 @duglin

duglin commented on api/client/run.go:
```diff
@@ -6,9 +6,11 @@ import (
 
 	"github.com/Sirupsen/logrus"
 	Cli "github.com/docker/docker/cli"
+	"github.com/docker/docker/errors"
```
Just for consistency, please alias this to "derr". since "errors" is used in other files for go's errors package.

duglin commented on api/client/run.go:
```diff
@@ -36,6 +38,23 @@ func (cid *cidFile) Write(id string) error {
+	}
+	return statusError
+}
+
```
Not your fault, but I can't tell you how much this code makes me shutter.   strcmp (and even worse, substr) on error messages stinks.

sallyom replied:
i agree...but i did my best to make it seem less string-comparey.

duglin commented on integration-cli/docker_cli_exec_test.go:
```diff
@@ -618,8 +617,8 @@ func (s *DockerSuite) TestExecStartFails(c *check.C) {
-	c.Assert(err, check.NotNil, check.Commentf(out))
-	c.Assert(out, checker.Contains, "executable file not found")
+	if !(err != nil || out == "" || strings.Contains(err.Error(), "Error running command in existing container")) {
+		c.Fatalf("exec with non-existent cmd should have errored out with no output")
```
Doesn't this seem wrong to show no output at all? Not even an "exe not found" type of error msg.
Also, shouldn't we be checking the exit code to make sure its correct?

sallyom replied:
when i run these 2 cmds:
docker run -d  --name exec-15750 busybox top
docker exec exec-15750 no-such-cmd
this spits out:
exec: "no-such-cmd": executable file not found in $PATH
but with dockerCmdWithError("exec", name, "no-such-cmd"), it's the err that contains "executable file not found in $PATH"  and out is ""
exitCode not checked because the container "exec-15750" still running...so it's 255 from the contained process
I may be confused, I've been staring at this for a long time...

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -3438,3 +3438,54 @@ func (s *DockerSuite) TestRunWrongCpusetMemsFlagValue(c *check.C) {
+	runCmd := exec.Command(dockerBinary, "run", "--name", name, "busybox", "foo")
+	_, exit, _ := runCommandWithOutput(runCmd)
+	stateExitCode := findContainerExitCode(c, name)
+	if !(exit == 127 && strings.Contains(stateExitCode, "127")) {
```
I like that you check both!!!

sallyom replied:
:) thnx

duglin commented on pkg/integration/utils.go:
```diff
@@ -27,6 +27,16 @@ func GetExitCode(err error) (int, error) {
 			return procExit.ExitStatus(), nil
 		}
 	}
+	// error if contained command not found/does not exist
```
Can you explain this code here? If we have to check the output error message to "calculate" the exit code, then isn't something wrong since I thought the point of this PR was to make sure our exit codes were set properly?

sallyom replied:
not ignoring this, just looking into it more to compare to derr.Errs...not very easy to test this package, actually.

sallyom replied:
@duglin as I (now) understand it the pkg/integration is to set up/test the utilities used in integration-cli, so there's no need to add tests or change anything there.  The new exit codes are tested in int-cli.  Thnx for your patience :)

duglin commented on pkg/integration/utils.go:
```diff
@@ -36,9 +46,8 @@ func ProcessExitCode(err error) (exitCode int) {
-			// we've failed to retrieve exit code, so we set it to 127
-			exitCode = 127
+			// we've failed to retrieve exit code, so we set it to 125
+			exitCode = 125
```
Might as well just make GetExitCode() return 125 in these cases instead of `exitCode, fmt.Errorf("failed to get....")` - then we can have it just return 'exitCode' w/o the error.  No one else calls GetExitCode() except right here anyway.

sallyom commented:
working on feedback now, I've fixed the 'image doesn't not exist' errors by adding a similar check in api/client/run.go for the already existing derr.ErrorCodeNoSuchImageHash and derr.ErrorCodeNoSuchImageTag errors that were created in daemon/create.go --should have seen this before :( working on other feedback now....

sallyom commented:
@LK4D4 @cpuguy83 I've addressed all feedback, thanks for giving it another look.
@duglin, see my response about the pkg/integration (no changes necessary) 
thanks :)

duglin commented on api/client/run.go:
```diff
@@ -199,8 +224,9 @@ func (cli *DockerCli) CmdRun(args ...string) error {
-	if _, _, err = readBody(cli.call("POST", "/containers/"+createResponse.ID+"/start", nil, nil)); err != nil {
-		return err
+	if _, _, err := readBody(cli.call("POST", "/containers/"+createResponse.ID+"/start", nil, nil)); err != nil {
+		cmd.ReportError(err.Error(), true)
```
I'm thinking the 2nd param should be `false` here.
```
$ docker run -ti ubuntu /tmp
exec: "/tmp": permission denied
docker: Error response from daemon: Contained command could not be invoked..
See '/root/docker/bundles/1.9.0-dev/binary/docker run --help'.
```
I think it makes sense to show the short help message in cases where we think there might be a docker specific error on the cmd line (like a bad option), but in cases where we've made it all the way to the container.start() we're probably into the user (or cmd) space and so I'm not sure showing the help message helps.  Keeping it on the parsing error and create error, above, are ok though.

duglin commented on errors/daemon.go:
```diff
@@ -872,4 +863,31 @@ var (
+	// ErrorCodeCantStart127 is generated when contained cmd can't start,
+	// contained command not found error, exit code 127
+	ErrorCodeCantStart127 = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "COMMANDNOTFOUND",
```
The pattern we're following is that the "Value" matches the ErrorCode, but in all caps. So this would be CANTSTART127

duglin commented:
overall looks good - just a couple of minor comments

sallyom commented:
@duglin, addressed your comments, thanks again.

duglin commented:
LGTM!

sallyom commented:
ping @LK4D4 @cpuguy83, anything else?  Thanks :)

bfirsh commented:
ping @jfrazelle @tiborvass look good?

calavera commented:
LGTM.

I'm rerunning the build before merging it to make sure this change doesn't break anything after so many changes.

sallyom commented:
@calavera anything else?

thaJeztah commented:
Hm, does this need documentation changes somewhere? I'm not sure actually if we mention exit-codes somewhere in the docs

thaJeztah commented:
I counted two LGTM's, moving to docs review (but I'm not actually sure we have something around this, so suggestions are welcome)

icecrime commented:
It's a good feature, I believe it should be documented indeed!

sallyom commented:
@thaJeztah @icecrime in man/docker-run.1.md and docs/reference/run.md?  I'm on it :)

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

thaJeztah commented on docs/reference/run.md:
```diff
@@ -518,6 +518,38 @@ non-zero exit status more than 10 times in a row Docker will abort trying to
 
+## Exit Status
+
+The exit code from `docker run` will give information about why the container
```
small nit: s/will give/gives/

thaJeztah commented:
> in man/docker-run.1.md and docs/reference/run.md? I'm on it :)

sgtm! I see you added to `reference/run.md`, not yet in the man page

looks like this needs to be squashed and rebased as well (don't shoot the messenger! :smile:)

sallyom commented:
@thaJeztah done!

calavera commented:
thanks @sallyom. Man LGTM too. @thaJeztah can you check it out and merge if it looks good to you?

thaJeztah commented:
Oh cool, it's updated, thanks!!

LGTM \o/

sallyom commented:
woohoo!

cpuguy83 commented:
/me waits for the world to burn with this one :)

This is a good change though, I like having distinct exit codes.

runcom commented:
@sallyom thanks! and great job!
